pkg/nsx/services/firewall.go

```diff
@@ -281,18 +363,209 @@ func (service *SecurityPolicyService) addOperatorIfNeeded(expressions *data.List
 	}
 }
 
-func (service *SecurityPolicyService) updatePortExpressions(matchLabels map[string]string, expressions *data.ListValue) {
+func (service *SecurityPolicyService) updateExpressionsMatchLables(matchLabels map[string]string, memberType string, expressions *data.ListValue) {
 	for k, v := range *util.NormalizeLabels(&matchLabels) {
 		service.addOperatorIfNeeded(expressions, "AND")
 		expression := service.buildExpression(
-			"Condition", "SegmentPort",
+			"Condition", memberType,
 			fmt.Sprintf("%s|%s", k, v),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(expression)
 	}
 }
 
+// NSX understand the multiple values w.r.t a key in a joined string manner
+// this function iterates over input matchExpressions LabelSelectorRequirement
+// with same operator and Key, and merges them into one and values to a joined string
+// e.g. {'NotIn': [{'k1': ['a1', 'a2', 'a3']}, {'k1': ['a2', 'a3', 'a4']}]}
+//   => {'NotIn': [{'k1': 'a1, a2, a3, a4'}]
+func (service *SecurityPolicyService) mergeSelectorMatchExpression(matchExpressions []metav1.LabelSelectorRequirement) *[]metav1.LabelSelectorRequirement {
+	mergedMatchExpressions := make([]metav1.LabelSelectorRequirement, 0)
+	isMerged := make([]bool, len(matchExpressions))
+	var meregedSelector metav1.LabelSelectorRequirement
+	var i, j = 0, 0
+	var isFound = false
+
+	for i = 0; i < len(matchExpressions); i++ {
+		if isMerged[i] {
+			continue
+		}
+		for j = i + 1; j < len(matchExpressions); j++ {
```

Comment:
Thanks for advice, I tried to the above example and maybe it's not exact we what we want. since we want to build union  Values set for the same operator and key value. the Data structure will have to  iterate each Values list of match expression list. but we can put Values in a set when  same operator and key value, using set union to create a merged set to remove duplicated  value from union set. it's will be easier and time complexity also be reduced since there is no nested-loop any more.